sanitizers/detectors/detector.go

```diff
@@ -1,9 +1,96 @@
 package detectors
 
-// Detector defines the interface for the bug detectors.
-type Detector interface {
-	// Detect performs checks to detect specific classes of bugs and returns specialized errors
-	// representing the found bugs. The function can also guide the fuzzer towards producing
-	// interesting inputs that trigger the bugs of interest.
-	Detect() error
+import (
+	"errors"
+	"fmt"
+	"os/exec"
+	"text/template/parse"
+
+	"github.com/CodeIntelligenceTesting/gofuzz/sanitizers/reporter"
+)
+
+type DetectorClass struct {
+	d         Detectors   // Concrete Detector for this instance
+	id        int         // Numeric identifier to distinguish between the detectors for the various call sites
+	cmd       string      // The tampered with SQL query/cmd/path/...
+	tree      *parse.Tree // Template injection specific field
+	err       error       // Generic error handler
+	extraArgs []any       // Any amount of extra parameters that are necessary for detection logic/report verbosity
```

Comment:
I'm concerned that combining all detectors in one struct would make the logic harder to understand and maintain. Here, `DetectorClass` uses fields that are only relevant for certain detectors, e.g., `cmd` and `tree`. This can become convoluted when we add further detectors that require new fields. I think we should adhere to the separation of concerns principle and keep our detectors as simple as possible. 

One source of code duplication here is that we check the returned error in the hooks and then report a finding if the error is of a certain type. I think it would be a good idea to move the reporting logic into the detectors. This way the hooks only need to call the `Detect` method and the detectors then take care of reporting the findings. This would remove the error-checking code from the hooks. WDYT?

Reply:
I concur for the most part. Currently, the logic is still fairly trivial so that I don't see much problems in combining all sanitizers into one struct, especially since the overlap between them is larger than the non-overlapping part. Once we work on expanding the current implementation I agree with the concerns though. It'd be favorable to have have one parent struct from which all childs inherit the properties they likely all share. This would be clean and reduce code-duplication for the most part.

Pushing the reporting back to the specific detectors is an easy feat. The current reporting mechanism is fairly slick for all cases so it wouldn't produce too long `Detect()` functions.

I'll give this some thought and push some changes after the holidays :)